Adding first pass at wayland docs (#1070)
AvaloniaUI/avalonia-docs: docs/platform-specific-guides/linux.md

---
id: linux
title: Desktop Linux
description: How Avalonia runs on desktop Linux, including the experimental Wayland backend, WSL 2 setup, and accessibility support with AT-SPI2.
doc-type: overview
---

## How Avalonia runs on Linux

Avalonia uses the Win32 API on Windows, its own native Objective-C++ backend on macOS, and on Linux it targets X11 by default. Most Linux distributions that support the .NET SDK and have X11, Wayland, or framebuffer capabilities will run Avalonia applications.

On Wayland desktops, Avalonia applications run through the XWayland compatibility layer by default. Starting with Avalonia 12.1.0, you can instead opt into the native [Wayland backend](#wayland).

## Wayland

The [`Avalonia.Wayland`](https://www.nuget.org/packages/Avalonia.Wayland) package provides a native Wayland backend. It communicates with the compositor using the Wayland protocol directly, instead of going through XWayland.

> The backend supports mouse, touch, and keyboard input, the clipboard, and drag-and-drop. Rendering uses OpenGL or OpenGL ES through EGL, with an optional dmabuf swapchain path.

Reviewer: ```suggestion
The backend supports mouse, touch, and keyboard input. It also supports clipboard and drag-and-drop. Rendering uses OpenGL or OpenGL ES through EGL, with an optional [dma-buf swapchain](https://docs.kernel.org/userspace-api/dma-buf-alloc-exchange.html) path.
```

> :::info

Reviewer: ```suggestion
:::caution
```

The Wayland backend is experimental. `UsePlatformDetect()` does not select it automatically; you must enable it explicitly.
:::

> ### Enable the Wayland backend

Reviewer: ```suggestion
### Enabling the Wayland backend
```

1. Add the `Avalonia.Wayland` package to your project:

   ```bash
   dotnet add package Avalonia.Wayland
   ```

2. Call `UseWayland()` on your [`AppBuilder`](/api/avalonia/appbuilder) in `Program.cs`:

   ```csharp
   public static AppBuilder BuildAvaloniaApp()
       => AppBuilder.Configure<App>()
           .UseWayland();
   ```

> `UseWayland()` always initializes the Wayland backend, and there is no automatic fallback: the application will not start in an environment without a Wayland compositor. For applications that also run on other operating systems or on X11 sessions, select the backend conditionally. For example, check the `WAYLAND_DISPLAY` environment variable that Wayland sessions set:

Reviewer: ```suggestion
`UseWayland()` always initializes the Wayland backend, and there is no automatic fallback: the application will not start in an environment without a Wayland compositor. For applications that must also run on other operating systems or on X11 sessions, select the backend conditionally. For example, check the `WAYLAND_DISPLAY` environment variable:
```

```csharp
public static AppBuilder BuildAvaloniaApp()
{
    var builder = AppBuilder.Configure<App>().UsePlatformDetect();

    if (OperatingSystem.IsLinux()
        && Environment.GetEnvironmentVariable("WAYLAND_DISPLAY") is not null)
    {
        builder = builder.UseWayland();
    }

    return builder;
}
```

### Configuration options

> To configure the backend, pass a `WaylandPlatformOptions` instance to the app builder:

Reviewer: ```suggestion
To configure the backend, pass a `WaylandPlatformOptions` instance to the `AppBuilder`:
```

```csharp
public static AppBuilder BuildAvaloniaApp()
    => AppBuilder.Configure<App>()
        .UseWayland()
        .With(new WaylandPlatformOptions
        {
            UseDmabufSwapchain = true
        });
```

| Option | Default | Description |
|---|---|---|
| `WlDisplayName` | `null` | The Wayland display to connect to (for example, `wayland-0`). When `null`, the `WAYLAND_DISPLAY` environment variable is used. |

> | `EnableReconnects` | Enabled | Reconnects to the compositor automatically when the connection is lost. |

Reviewer: ```suggestion
| `EnableReconnects` | `true` | Reconnects to the compositor automatically when the connection is lost. |
```

> | `UseDmabufSwapchain` | `null` | Uses a dmabuf-based swapchain for GPU rendering. When `null`, the backend decides based on compositor and driver capabilities. |

Reviewer: ```suggestion
| `UseDmabufSwapchain` | `null` | Uses a dma-buf-based swapchain for GPU rendering. When `null`, the backend decides based on compositor and driver capabilities. |
```
